unsafe/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -380,7 +405,13 @@ public void putNewKey(
       bitset.set(pos);
 
       // If there's not enough space in the current page, allocate a new page:
-      if (currentDataPage == null || PAGE_SIZE_BYTES - pageCursor < requiredSize) {
+      if (currentDataPage == null || PAGE_SIZE_BYTES - pageCursor - 8 < requiredSize) {
```

Comment:
There's an assertion on line 403 that needs to be updated to reserve 8 extra bytes for the final length.